### PR TITLE
 Account for type params on method without parentheses

### PR DESCRIPTION
**Pull request by estebank**
Account for those type parameters in the structured suggestion when forgetting to call method:

```
error[E0615]: attempted to take value of method `collect` on type `std::vec::IntoIter<_>`
  --> $DIR/method-missing-parentheses.rs:2:32
   |
LL |     let _ = vec![].into_iter().collect::<usize>;
   |                                ^^^^^^^---------
   |                                |
   |                                help: use parentheses to call the method: `collect::<usize>()`
```

### CODE REVIEW

**Comment by estebank** [comment]
The second commit can be reverted if it is too intrusive for too small a gain.

**Comment by petrochenkov** [comment]
Looks like the second comment also makes fields with generic arguments legal under `#[cfg(FALSE)]`, which is a language change.

**Comment by estebank** [comment]
@petrochenkov adding delay_span_bug would help with accidental stabilization, but not be a good experience. Is there a place where this check could be done even if cfged out?

File: src/librustc_typeck/check/expr.rs [hunk 1]
```diff
@@ -1440,6 +1440,15 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         display
     }
 
+    fn complain_about_invalid_type_params(&self, args: Option<Span>) {
```
Comment (Centril) [inline src/librustc_typeck/check/expr.rs]: Let's move this to after `check_field` as the other error methods related to `check_field` also come after.

Comment (Centril) [inline src/librustc_typeck/check/expr.rs]: Please add tests for each situation in which this method is called so the method is exercised.

File: src/librustc_typeck/check/expr.rs [hunk 2]
```diff
@@ -1440,6 +1440,15 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+        if let Some(args) = args {
+            self.tcx
+                .sess
+                .struct_span_err(args, "field expressions may not have generic arguments")
```
Comment (Centril) [inline src/librustc_typeck/check/expr.rs]:
```suggestion
                .struct_span_err(args, "field expressions cannot have generic arguments")
```

File: src/librustc_parse/parser/expr.rs [hunk 3]
```diff
@@ -825,16 +825,12 @@ impl<'a> Parser<'a> {
             Ok(self.mk_expr(span, ExprKind::MethodCall(segment, args), AttrVec::new()))
         } else {
             // Field access `expr.f`
-            if let Some(args) = segment.args {
-                self.struct_span_err(
-                    args.span(),
-                    "field expressions may not have generic arguments",
-                )
-                .emit();
```
Comment (Centril) [inline src/librustc_parse/parser/expr.rs]: I think your only recourse beyond doing the language change is to `.stash` the error instead and steal it in typeck. (And then add tests demonstrating that it still fails under `cfg(FALSE)`.)

Comment (petrochenkov) [reply src/librustc_parse/parser/expr.rs]: Please, no.
Just remove the second commit.
It's to much garbage for nearly zero benefit.

Comment (Centril) [reply src/librustc_parse/parser/expr.rs]: (I'm just laying out the options; I don't have a particular opinion.)

Comment (estebank) [reply src/librustc_parse/parser/expr.rs]: removed the commit

Comment (estebank) [reply src/librustc_parse/parser/expr.rs]: removed the commit

**Review by davidtwco — APPROVED** [review]

**Pull request merged**